src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp

```diff
@@ -126,173 +127,187 @@ inline void PSPromotionManager::push_contents(oop obj) {
     obj->oop_iterate_backwards(&pcc);
   }
 }
+
+template<bool promote_immediately>
+inline oop PSPromotionManager::copy_to_survivor_space(oop o) {
+  assert(should_scavenge(&o), "Sanity");
+
+  // NOTE! We must be very careful with any methods that access the mark
+  // in o. There may be multiple threads racing on it, and it may be forwarded
+  // at any time.
+  markWord m = o->mark();
+  if (!m.is_marked()) {
+    return copy_unmarked_to_survivor_space<promote_immediately>(o, m);
+  } else {
+    // Ensure any loads from the forwardee follow all changes that precede
+    // the release-cmpxchg that performed the forwarding, possibly in some
+    // other thread.
+    OrderAccess::acquire();
```

Comment:
Maybe add a comment here that `copy_unmarked_to_survivor_space` also guarantees this for all paths (i.e. both successful as well as failing CAS). Maybe this is not the correct place here, maybe in a comment for `copy_unmarked_to_survivor_space`; or something like "`copy_to_survivor_space` acts as an acquire" or something, so that the returned object can be used "safely" - idk.